testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprContainer.java

```diff
@@ -0,0 +1,330 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package io.dapr.testcontainers;
+
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.images.builder.Transferable;
+import org.testcontainers.utility.DockerImageName;
+import org.yaml.snakeyaml.DumperOptions;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.nodes.Tag;
+import org.yaml.snakeyaml.representer.Representer;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DaprContainer extends GenericContainer<DaprContainer> {
+
+  private static final int DAPRD_HTTP_PORT = 3500;
+  private static final int DAPRD_GRPC_PORT = 50001;
+  private final Set<Component> components = new HashSet<>();
+  private final Set<Subscription> subscriptions = new HashSet<>();
+  private String appName;
+  private Integer appPort = 8080;
+  private DaprLogLevel daprLogLevel = DaprLogLevel.INFO;
+  private String appChannelAddress = "localhost";
+  private String placementService = "placement";
+  private static final DockerImageName DEFAULT_IMAGE_NAME = DockerImageName.parse("daprio/daprd");
+  private Yaml yaml;
+  private DaprPlacementContainer placementContainer;
+  private String placementDockerImageName = "daprio/placement";
+  private boolean shouldReusePlacement;
+
+  /**
+   * Creates a new Dapr container.
+   * @param dockerImageName Docker image name.
+   */
+  public DaprContainer(DockerImageName dockerImageName) {
+    super(dockerImageName);
+    dockerImageName.assertCompatibleWith(DEFAULT_IMAGE_NAME);
+    // For susbcriptions the container needs to access the app channel
+    withAccessToHost(true);
+    // Here we don't want to wait for the Dapr sidecar to be ready, as the sidecar
+    // needs to
+    // connect with the application for susbcriptions
+
+    withExposedPorts(DAPRD_HTTP_PORT, DAPRD_GRPC_PORT);
+
+    DumperOptions options = new DumperOptions();
+    options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+    options.setPrettyFlow(true);
+
+    Representer representer = new YamlRepresenter(options);
+    representer.addClassTag(MetadataEntry.class, Tag.MAP);
+    this.yaml = new Yaml(representer);
```

Comment:
Please, make this.yaml as final and initialize from a static method call in the top of this method.